### ISSUE TITLE
Electron/div based title bar

### ISSUE THREAD

**Issue by Daeraxa** [pulsar-edit/pulsar#345]
### Have you checked for existing feature requests?

- [X] Completed

### Summary

Currently Pulsar (at least on Windows and Linux) by default uses the system title bar which is perfectly functional but:
1) Looks a bit dated
2) (On Windows at least) has its colour based on the current light/dark theme in the OS, not in Pulsar.
3) Can't really be customised
4) Takes up a lot of vertical space which cannot be shrunk.

Pulsar already includes the ability to have a "hidden" title bar but this simply removes it entirely. Instead of relying on the native system one we could instead create one as part of the actual application as an option.

macOS doesn't have a "Window" titlebar but an option could be provided for macOS that still allows the same customisation but simply omits the menus that would be on the system menu bar instead, still providing access to any "new" buttons or items such as a settings "cog" or any info type displays.

### What benefits does this feature provide?

- Allows us to customise and modify the titlebar as we want which could include extra features, buttons, icons etc. without being constrained by the OS
- Provides a more "modern" look which we can style along with the rest of the application as it would simply be another div (or similar style-able element).
- Allows the title bar to correctly follow the application, rather than the OS, theme.
- Provides possibility for community packages to add to or modify data in the title bar.
- Could move items currently occupying the bar at the bottom on Pulsar to the top instead to provide some additional vertical space on the window.

### Any alternatives?

It seems somebody has indeed made a package for this in the past - https://web.pulsar-edit.dev/packages/atom-windows-titlebar but if we were to do this in the core application we could standardise it easier for package developers and themes to take into account.

### Other examples:

VSCode does this by default and therefore has additional options for the pane layouts:

![image](https://user-images.githubusercontent.com/58074586/214606280-6cc7b128-1682-4e6d-9f24-52fca285fd7c.png)

This is also common on other Electron applications (e.g. Discord).

**Comment by confused-Techie** [comment]
Interesting to see that we could have something to pass this off of.

Although do we want to add it as another Core Package? Would there be any advantage to baking it right into the editor?

Since I think even a package can be extensible to other packages right?

**Comment by Daeraxa** [comment]
I don't see it as a core package so much as just part of the editor source directly, potentially with a view to making it the default over using the OS menu bar eventually. Having it as an optional package would mean people probably not noticing it or adding support for it into new packages.

**Comment by confused-Techie** [comment]
@Daeraxa you make a good point there. Just thought it could be core since even the tabs are a core package. 

But Also totally open for it being put right into the core of the application then exposing it as an API. So sounds good to me

**Comment by autumnblazey** [comment]
also, on macos, "custom titlebar" in vscode is available, but it doesn't contain any of the menubar items since that is still handled by the OS (though the layout and command centre and other stuff are still available to be put in the title bar)

<img width="628" alt="image" src="https://user-images.githubusercontent.com/49562048/214734706-da220f4a-533d-445a-86b2-40837473d122.png">

**Comment by Daeraxa** [comment]
> macOS doesn't have a "Window" titlebar but an option could be provided for macOS that still allows the same customisation but simply omits the menus that would be on the system menu bar instead, still providing access to any "new" buttons or items such as a settings "cog" or any info type displays.

Yeah, this is what I was trying to get across with this bit:
> macOS doesn't have a "Window" titlebar but an option could be provided for macOS that still allows the same customisation but simply omits the menus that would be on the system menu bar instead, still providing access to any "new" buttons or items such as a settings "cog" or any info type displays.

> you make a good point there. Just thought it could be core since even the tabs are a core package.
> But Also totally open for it being put right into the core of the application then exposing it as an API. So sounds good to me

So my main thought here was that as a package, even a core one, it loads after the main editor which just seems a bit weird for something as fundamental as the title bar of the application.

Actually I'm wondering if macOS is *already* using this, just not in a position right now to look into it.
There is already a `.title-bar` selector and stuff in src/title-bar.js which includes code for maximising/minimising the application so it looks like the framework for this already exists and we might just be able to extend it. I should be able to have a proper look this weekend but it might be that we don't even need to re-invent the wheel here.

**Comment by savetheclocktower** [comment]
> Actually I'm wondering if macOS is _already_ using this, just not in a position right now to look into it.
> There is already a `.title-bar` selector and stuff in src/title-bar.js which includes code for maximising/minimising the application so it looks like the framework for this already exists and we might just be able to extend it. I should be able to have a proper look this weekend but it might be that we don't even need to re-invent the wheel here.

Yeah, I've been using this for years. It's the easiest way to ensure your title bar harmonizes with your theme. (There's a specific `@title-bar-background-color` UI variable that defaults to `@base-background-color`.) But it doesn't try to do anything radical. I suspect one of the reasons why it's only an option on macOS is that it avoids any questions about how to handle the menu bar (since on macOS it's always at the top of the screen).

If we wanted to play around with this a bit more, step one would be to go into `config-schema.js` and change the `core.titleBar` setting so that the `custom` and `custom-inset` options are available on all platforms.
